### PR TITLE
Support fetching remote templates

### LINKED ISSUE #233: Fetch template dependencies on the fly
We can borrow the idea that Bazel implements dependencies, e.g. `http_archive`, `http_file`, `git_repository` https://github.com/bazelbuild/bazel/blob/master/WORKSPACE#L378 and we can download the templates on the fly at the beginning of engine and policygen execution.

### PR DESCRIPTION
**Pull request by umairidris**
#233

### CODE REVIEW

File: cmd/tfengine/main.go [hunk 1]
```diff
@@ -41,7 +42,8 @@ func main() {
 		log.Fatal("--output_path must be set")
 	}
 
-	if err := tfengine.Run(*configPath, *outputPath, *format); err != nil {
+	opts := &tfengine.Options{Format: *format, CacheDir: *cacheDir}
+	if err := tfengine.Run(*configPath, *outputPath, opts); err != nil {
```
[the remark this answers is not in the document]
Comment (umairidris) [reply cmd/tfengine/main.go:46]: Options are for optional things that can have a default value. If this was a user facing API I would have implemented these as functional options: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis but we don't need to do that right away as the values are coming from flags.

File: internal/tfengine/tfengine.go [hunk 2]
```diff
@@ -136,25 +141,52 @@ func dumpTemplate(conf *Config, root, outputPath string, ti *templateInfo) error
...
+		}
+		if !filepath.IsAbs(path) {
+			path = filepath.Join(pwd, path)
+		}
+		return path, nil
+
+	}
+	dst := filepath.Join(cacheDir, "templates")
+	root, subdir := getter.SourceDirSubdir(path)
+	c := getter.Client{
```
Comment (xingao267) [inline internal/tfengine/tfengine.go:178]: this is neat

File: internal/tfengine/tfengine.go [hunk 3]
```diff
@@ -136,25 +141,52 @@ func dumpTemplate(conf *Config, root, outputPath string, ti *templateInfo) error
...
 	}
 	return nil
 }
 
+// fetchPath handles fetching remote paths.
+// If the path is local then it will simply expand it and return.
+// Remote paths are fetched to the cache dir.
+// Relative paths are handled from pwd.
+func fetchPath(path, pwd, cacheDir string) (string, error) {
+	if strings.HasPrefix(path, ".") { // Is local path.
```
Comment (xingao267) [inline internal/tfengine/tfengine.go:165]: what if the local path is an absolute path?

File: internal/tfengine/tfengine.go [hunk 4]
```diff
@@ -136,25 +141,52 @@ func dumpTemplate(conf *Config, root, outputPath string, ti *templateInfo) error
...
-			cp = filepath.Join(root, cp)
-		}
 		if err := template.WriteDir(cp, outputPath, data); err != nil {
 			return fmt.Errorf("component %q: %v", cp, err)
 		}
 	}
 	return nil
 }
 
+// fetchPath handles fetching remote paths.
```
Comment (xingao267) [inline internal/tfengine/tfengine.go:160]: move this to a common package so policygen can use it

Comment (xingao267) [reply internal/tfengine/tfengine.go:160]: maybe pathutil

File: cmd/tfengine/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MartinPetkov) [reply cmd/tfengine/main.go:46]: I see, thanks for the explanation.

File: internal/tfengine/tfengine.go (hunk 3 above)
Comment (MartinPetkov) [reply internal/tfengine/tfengine.go:165]: Maybe Stat the string and check if it's a file path.

Comment (umairidris) [reply internal/tfengine/tfengine.go:165]: Absolute paths are not currently supported and involves some care to implement. I based the current logic on https://github.com/hashicorp/terraform/blob/807267d1b56aa6298fecf5edba98c088c2866e83/internal/initwd/getter.go#L177. Terraform does some further checks for absolute paths I think which I can port later on.

Comment (umairidris) [reply internal/tfengine/tfengine.go:165]: stat seems like a good idea, I can do that

Comment (umairidris) [reply internal/tfengine/tfengine.go:165]: os.Stat isn't working out too well either, I'll revisit this later.

**Comment by xingao267** [comment]
#233 includes both tfengine and policygen. So maybe leave it open after this is merged.

**Comment by umairidris** [comment]
> #233 includes both tfengine and policygen. So maybe leave it open after this is merged.

got it

**Pull request merged**